rfcs/0003-image-filter.md

```diff
@@ -0,0 +1,176 @@
+    * https://github.com/sampeinado/kpack/blob/master/rfcs/0002-reason-message-diff.md
+
+* Issue that stemmed from the user interview with Jason Morgan in which he requested the ability to filter the `kp build list` table https://github.com/vmware-tanzu/kpack-cli/issues/79 
+* Kubernetes Field Selectors https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/
+
+### Complexity/Risk
+
+* By introducing a new column in `kp image list`, we may encounter a similar width problem as we did with the builds list table
+
+### Alternatives:
+
+* Users parse their image list as it currently exists and scroll through it until they find the images with the attributes they care about
+* Usage of kubectl to return images with certain attributes
+
+### Outcome:
+
+Propose an additional feature or set of features that enable users to query kpack images based on their `status` and `latest-reason` such that it is easier for users to perform the workflows in the “Use Cases” section.  In order to make this filtration feature more effective, users should be able to list all images in the cluster (dependent on the user's RBAC permissions). As an implication of this change, the table should include a `NAMESPACE` column when the user filters for images in all namespaces.  Following a round of feedback during a working group meeting on 10/9, I am additionally proposing the inclusion of `builder` and `clusterbuilder` flags.  This modification addresses the concern that that we expect kpack images to be updating frequently, thereby making it difficult for the end-user to verify a stack or store update if the images have updated more recently based on different `REASON`.  These flags would effectively enable the user to display the last time and image was rebuilt for `Reason` `BUILDPACK` or `STACK` 
+
+Although the feature request stated in the “Problem” section is to query builds, it is probably more productive to query images because end-users most likely care about the current state of their application build and not builds that happened in the past.  Additionally, querying based on builds would pose a technical problem as the CLI would need to present builds and their build numbers outside the context of their associated image.  
+
+
+### Mockups
+
+Below I will cover specific scenarios made possible by a filtering feature. All the following scenarios are derived from this proposed modification to the `kp image list` command:
+
+```
+kp image list --status <string1, string 2 ...> --latest-reason <string1, string2 ...>
+```
+
+Rather than adding an additional subcommand to the `kp image list` path like `filter` or `query`, I propose adding an additional `LATEST REASON` column to the existing output of the `kp image list` table. This would also allow for future design extension. If we wanted to add additional filtering capabilities, we would just create new attribute flags with values as arguments.
```

Comment:
I believe that using an individual flag for each type of filter could lead to an explosion of possible filter flags. Additionally, using a filter or query syntax is pretty common in CLIs (thats what kubectl does) and allows for more extensibility later on (via query syntax improvements).

Reply:
I kinda like the ui simplicity of individual flags, but if the possible filtering parameters is too high I think it becomes not worth the simplicity. We could also potentially support both...

Reply:
I like the idea of just a filter flag like kubectl. I feel like it's pretty clean and easy to use from a user standpoint.

Reply:
One thing that confuses me about this suggestion is that it seems like we'd be trading the complexity of flags for the complexity of arguments.  The user still has to specify the values they want to filter.  So if we did something like `kp image filter/query` we'd have to support a list of arguments to be used.  Is that less technically complex than supporting a bunch of flags?  I also slightly prefer the UX of the current implementation because it can be reconciled with `kp image list`.  The user is effectively trying to filter this list, not generate their own list based on inputs (which is effectively what `kp image filter` would be.  @tomkennedy513 @tylerphelan @djoyahoy lmk your thoughts.

Reply:
I don't know if it's less technically complex to maintain, but I think having a filter query flag vs individual flags would accomplish the same thing. I am a little confused by the comment that users are not trying to generate their own list mainly because my understanding of the filter query method would still return all columns (i.e not creating some new data view) so if a user ran something like `kp image list --lastest-reason "config"` vs `kp image list --filter latest-reason=config` they would get the same result. I prefer the filter flag because it is how `kubectl` does it and since users are likely using both `kp` and `kubectl` together, I figure it would be slightly less cognitive load if the  filtering mechanism was identical

Reply:
I agree with @djoyahoy and @tomkennedy513 that we should follow the kubectl convention and use a filter flag. We should keep in mind Jakob's Law of the Internet User Experience: "Users spend most of their time on other sites." This means that users prefer your tool to work the same way as all the other tools they already know.

Reply:
As far as I can tell, the flag for filtering a `kubectl get` command is `--selector` (`-l`) or `--field-selector` but I can't quite tell what the difference between them is

Reply:
@mgibson1121 i dont think anyone is proposing a new subcommand of `kp image filter` or something like that. the proposal would be `kp image list --selector reason==config". Am I right @djoyahoy @tomkennedy513 ?

Reply:
This strikes me as a different case than Kubectl because it just runs matching on keys and values for labels (these labels are frequently managed by end-users of kubectl).  So this functionality strikes me as a generic implementation so that they can make something generically useful for lots of users.  In this RFC we're trying to find an interface to allow for filtering of an opinionated set of criteria.   Based on this feedback and my above response, I propose the following syntax as a path forward: `kp image list --latest-reason=value1, value2`.  We would provide help text for values that are supported.

Reply:
after talking about this with Matt, I also feel like this is the best path forward

Reply:
@tomkennedy513 Could you add a little info on your thought process/conversation there?

Reply:
We're going with this structure.  I will update the mockups to reflect this decision. 

`kp image list --filter reason=stack, config --field-selector clusterbuilder=value`